bitswap/network/httpnet/httpnet.go

```diff
@@ -0,0 +1,378 @@
+package httpnet
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	bsmsg "github.com/ipfs/boxo/bitswap/message"
+	pb "github.com/ipfs/boxo/bitswap/message/pb"
+	"github.com/ipfs/boxo/bitswap/network"
+	blocks "github.com/ipfs/go-block-format"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	probing "github.com/prometheus-community/pro-bing"
+)
+
+var log = logging.Logger("httpnet")
+
+var ( // todo
+	maxSendTimeout = 2 * time.Minute
+	minSendTimeout = 10 * time.Second
+	sendLatency    = 2 * time.Second
+	minSendRate    = (100 * 1000) / 8 // 100kbit/s
+)
+
+var ErrNoHTTPAddresses = errors.New("AddrInfo does not contain any valid HTTP addresses")
+
+var _ network.BitSwapNetwork = (*httpnet)(nil)
+
+type Option func(net *httpnet)
+
+type httpnet struct {
+	// NOTE: Stats must be at the top of the heap allocation to ensure 64bit
+	// alignment.
+	stats network.Stats
+
+	host   host.Host
+	client *http.Client
+
+	// inbound messages from the network are forwarded to the receiver
+	receivers []network.Receiver
+
+	urlLock   sync.RWMutex
+	peerToURL map[peer.ID][]*url.URL
+	urlToPeer map[string]peer.ID
+
+	latMapLock sync.RWMutex
+	latMap     map[peer.ID]time.Duration
+}
+
+// New returns a BitSwapNetwork supported by underlying IPFS host.
+func New(host host.Host, opts ...Option) network.BitSwapNetwork {
+	dialer := &net.Dialer{
+		Timeout:   30 * time.Second,
+		KeepAlive: 30 * time.Second,
+	}
+	c := &http.Client{
+		Transport: &http.Transport{
+			Proxy:                 http.ProxyFromEnvironment,
+			DialContext:           dialer.DialContext, // maybe breaks wasm
+			ForceAttemptHTTP2:     true,
+			MaxIdleConns:          100,
+			IdleConnTimeout:       90 * time.Second,
+			TLSHandshakeTimeout:   10 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+		},
+	}
+
+	net := httpnet{
+		host:      host,
+		client:    c,
+		peerToURL: make(map[peer.ID][]*url.URL),
+		urlToPeer: make(map[string]peer.ID),
+		latMap:    make(map[peer.ID]time.Duration),
+	}
+
+	for _, opt := range opts {
+		opt(&net)
+	}
+
+	return &net
+}
+
+func (bsnet *httpnet) Start(receivers ...network.Receiver) {
+	bsnet.receivers = receivers
+}
+
+func (bsnet *httpnet) Stop() {
+}
+
+func (bsnet *httpnet) Ping(ctx context.Context, p peer.ID) ping.Result {
+	log.Debugf("Ping: %s", p)
+
+	pi := bsnet.host.Peerstore().PeerInfo(p)
+	urls := network.ExtractURLsFromPeer(pi)
+	if len(urls) == 0 {
+		return ping.Result{
+			Error: ErrNoHTTPAddresses,
+		}
+	}
+
+	// pick the first one. In general there should not be more than one
+	// url per peer. FIXME: right?
+	pingURL := urls[0]
```

Comment:
I'd expect dns resolution chooses which one to try in that case... but yeah we should try what we have. I would not ping though, just try to make the request directly, probably randomizing the list order.

Reply:
I thought this code was somewhere else. IIRC ping is only used to record latency to prioritize...

The pinger code prioritizes ipv4 over ipv6 and chooses the first resolved-entry... I think it may be a bit too much to consider hosts that have significantly different latencies depending on ipv4 or 6.

Ultimately the question is : given a provider URL `example.com`. How many providers is that? One or N?

i.e. example.com can resolve to multiple ipv4/ipv6s. Are each of them meant to be a different provider URL to which we trigger requests? or do we assume the request should go to the first result?

I think we should consider it a single provider. If they want to be considered separate providers they can by providing on different urls.